File: dialogio/Dialog_SubPos.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static void Dialog_SubPos (uint16_t *p)
{
  uint16_t t = *p;

  if ((t > 1) && _dlg.dbuf[t - 1 - 1].use)
    t--;
  else if (t > 1)
  {
    t--;
    Dialog_SubPos (&t);
  }
  if (_dlg.dbuf[t - 1].use)
    *p = t;
}

File: dialogio/pas/Dialog_SubPos.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

// Local to `Dialog' procedure

procedure SubPos(var p: Word);

var
  temp: Word;

begin
  temp := p;
  If (temp > 1) and dbuf[temp-1].use then Dec(temp)
  else If (temp > 1) then begin Dec(temp); SubPos(temp); end;
  If dbuf[temp].use then p := temp;
end;
